tests/unittest.py

```diff
@@ -174,6 +173,7 @@ def create_peer_from_builder(self, builder, start_manager=True):
 
         if start_manager:
             manager.start()
+            self.clock.running = True
```

Comment:
Why do you have to manually mark it as running?

Reply:
Because before we were using a `Clock` as a reactor for tests, and `Clock` does not implement the `IReactorCore` interface, which it should since this interface's methods are actually used by us. More specifically, `IReactorCore` defines the `run()` method. It also defines the `running` attribute that is being used by our current PubSub implementation (it ignores the attribute when testing since `Clock` doesn't implement it).

By changing our reactor tests to the `MemoryReactorClock`, which does implement `IReactorCore`, the `run()` method and `running` attribute are defined. The ideal then would be to simply call `self.clock.run()`, but `MemoryReactorClock`'s fake `run()` implementation [weirdly calls `stop()`](https://github.com/twisted/twisted/blob/39ee213fca02e08bc1aadad61607d7e7e123a776/src/twisted/internet/testing.py#L516) before returning, making it unusable to us, since we need the reactor in a `running=True` state for the tests. For that, instead of calling `run()`, we manually set the attribute value.

Another option would be to subclass `MemoryReactorClock` and override our own `run()` implementation, but since the only thing it would do would be to set `running=True`, I thought it seemed a bit overkill.

Reply:
I agree it's very weird to have the `run()` method calling `stop()`. But, with the current implementation, we have to remember this fact in the future. So I think we should subclass the clock and override the `run()` implementation (and leave a note about this weird behavior from the original class).

Reply:
Fair enough, done in https://github.com/HathorNetwork/hathor-core/pull/694/commits/aa53857b5974f4082741ced56e763fb75de54e3d